### PR TITLE
Add logic to push prometheus metrics on each execution

### PR DESCRIPTION
**Pull request by dwelch0**
This PR adds two custom metrics `vault_manager_reconcile_success` and `vault_manager_duration_seconds` (followed qr metric naming format of `qontract_reconcile_*`). These metrics are exported to a push gateway on each execution of vault manager.

`vault_manager_reconcile_success` includes a label for `vault_instance`. This will allow for granular severity alerting when reconciliation of various vault instances fail.

Note: a flag `metrics` was included (defaults to false) so that metric exporting can be avoided within unit tests, environments that do not utilize push gateway, local executions, etc.

Result of testing against stage push gateway:
image

### CODE REVIEW

**Review by esron — APPROVED** [review]

**Review by cubismod — APPROVED** [review]
LGTM

File: pkg/utils/metrics.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	RECONCILE_SUCCESS_METRIC = "vault_manager_reconcile_success"
+	DURATION_METRIC          = "vault_manager_duration_seconds"
+)
+
+// push new values to vault_manager_reconcile_success for each instance reconciled
+func PushInstanceReconcileMetric(pushGatewayUrl string, instanceSuccess map[string]int) error {
+	vaultReconcileSuccessGauge := prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: RECONCILE_SUCCESS_METRIC,
+			Help: `Whether or not last reconcile was successful. ` +
```
Comment (bkez322) [inline pkg/utils/metrics.go:22]:
```suggestion
			Help: `Whether or not last reconcile for a particular vault instance was successful. ` +
```

File: pkg/utils/metrics.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	return nil
+}
+
+// push new total execution time metric
+func PushExecutionDurationMetric(pushGatewayUrl string, duration time.Duration) error {
+	executionDurationGauge := prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: DURATION_METRIC,
+			Help: "Execution duration of this job in seconds.",
```
Comment (bkez322) [inline pkg/utils/metrics.go:47]:
```suggestion
			Help: "Execution duration of this job (reconciling all vault instances) in seconds.",
```

**Review by bkez322 — CHANGES_REQUESTED** [review]

**Comment by bkez322** [comment]
Two minor suggestions from me but other LGTM.

**Comment by geoberle** [comment]
q-r generally exposes metrics regularly via a scrape endpoint and gets scraped via a service monitor. Generally Push gateway has its quirks and should be used only if regular scraping is not an option. What is the reason you decided to use the push gateway for this integration?

**Comment by rporres** [comment]
I do agree with @geoberle in avoiding the use of Push Gateway as much as possible as it works sometimes unexpectedly and it has a very complicated HA story (we don't have it deployed in HA). Isn't it possible to have a `metrics` endpoint in this integration the same way we have it in the q-r ones?

**Comment by geoberle** [comment]
@dwelch0 understood. vault-manager is not running as a daemon but within a bash loop that keeps the container alive while running the vault-manager binary occasionally. so no http-server would be running that could be scraped. did i get that right.

what do you think, would it be an option to put the while loop within the vault-manger itself? that would allow you to keep the process running and offering an http-server to scrape?

**Comment by dwelch0** [comment]
@geoberle following our slack discussion. Your suggestion on moving the execution loop to main (allowing for a http server to be persisted) seems the clear choice. I will revert this change.

**Pull request merged**